pennylane/io/qualtran_io.py

```diff
@@ -418,3 +584,358 @@ def compute_matrix(*params, **hyperparams):  # pylint: disable=no-self-argument
+    array([[1.+0.j, 0.+0.j, 0.+0.j, 0.+0.j],
+    [0.+0.j, 1.+0.j, 0.+0.j, 0.+0.j],
+    [0.+0.j, 0.+0.j, 0.+0.j, 1.+0.j],
+    [0.+0.j, 0.+0.j, 1.+0.j, 0.+0.j]])
+    """
+
+    _error_message = (
+        "Optional dependency 'qualtran' is required "
+        "for ToBloq functionality but is not installed. Try `pip install qualtran`."
+    )
+
+    def __init__(self, op, map_ops=False, **kwargs):
+        if not qualtran:
+            raise ImportError(self._error_message)
+
+        if not isinstance(op, Operator) and not isinstance(op, QNode):
+            raise TypeError(f"Input must be either an instance of {Operator} or {QNode}.")
+
+        self.op = op
+        self.map_ops = map_ops
+        self._kwargs = kwargs
+        super().__init__()
+
+    @cached_property
+    def signature(self) -> "qt.Signature":
+        """Compute and return Qualtran signature for given op or QNode."""
+        if isinstance(self.op, QNode):
+            self.op.name = "QNode"
+            num_wires = len(construct_tape(self.op)(**self._kwargs).wires)
+        else:
+            num_wires = len(self.op.wires)
+        return qt.Signature([qt.Register("qubits", qt.QBit(), shape=num_wires)])
+
+    def decompose_bloq(self):  # pylint:disable=too-many-branches
+        """Decompose the bloq using the op's decomposition or the tape of the QNode"""
+        try:
+            if isinstance(self.op, QNode):
+                tape = construct_tape(self.op)(**self._kwargs)
+                ops = tape.circuit
+                all_wires = list(tape.wires)
+            else:
+                ops = self.op.decomposition()
+                all_wires = list(self.op.wires)
+
+            signature = self.signature
+            in_quregs = out_quregs = {"qubits": np.array(all_wires).reshape(len(all_wires), 1)}
+
+            in_key = list(in_quregs.keys())[0]
+            out_key = list(out_quregs.keys())[0]
+
+            in_quregs = {
+                in_key: np.apply_along_axis(
+                    _QReg, -1, in_quregs[in_key], signature.get_left(in_key).dtype
+                )
+            }
+            out_quregs = {
+                out_key: np.apply_along_axis(
+                    _QReg, -1, out_quregs[out_key], signature.get_right(out_key).dtype
+                )
+            }
+
+            bb, initial_soqs = qt.BloqBuilder.from_signature(signature, add_registers_allowed=False)
+
+            # `signature.lefts()` can be thought of as input qubits. For our purposes LEFT and
+            # RIGHT signatures will (almost) always match since there are no allocated & freed
+            # qubits. Here, qreg_to_qvar is a map between a register and a Soquet. This serves
+            # as the foundation to wire up the rest of the bloqs.
+            qreg_to_qvar = {}
+            for reg in signature.lefts():
+                assert reg.name in in_quregs
+                soqs = initial_soqs[reg.name]
+                assert in_quregs[reg.name].shape == soqs.shape
```

Comment:
Change asserts to `ValueError`s for better error management, unless there's some special blocker.

Reply:
If I write them as ValueErrors, I cannot cover these with codecov, because there is no way for me to trigger the error. In other words, if a ToBloq(op).decompose_bloq() somehow raised these errors, it is not due to the input, but rather that the function itself is bugged.

I have these asserts so that if/when decompose_bloq() is updated (for example, to handle the new decompositions), we can rely on these asserts to tell us if there is a bug in the logic. Because the logic for translating decompositions is rather convoluted, keeping these asserts offers more value than removing them.